Remove multiraft; access etcd/raft directly from storage

bdarnell opened the pull request:
As described in docs/rfcs/dismantle_multiraft.md, this simplifies
concurrency and removes layers of abstraction that have not proven
helpful.

This commit contains two deliberate regressions: coalesced heartbeats
are not supported and raft groups are not created lazily. Both features
will be reintroduced in future commits.

I apologize for the size of this change but there were no good working intermediate states. At least it's a net reduction in code size and a lot of the churn was in a few files that were moved with only minor changes (`storage/raft.{go,pb.go,proto}` and `raft_transport.go`)

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3528)

<!-- Reviewable:end -->

bdarnell committed:
Bypass multiraft; access etcd/raft directly from storage.

As described in docs/rfcs/dismantle_multiraft.md, this simplifies
concurrency and removes layers of abstraction that have not proven
helpful.

This commit contains two deliberate regressions: coalesced heartbeats
are not supported and raft groups are not created lazily. Both features
will be reintroduced in future commits.

bdarnell committed:
Remove package multiraft.

BramGruneir commented:
Wonderful.  I'll start the code review tonight.

tschottdorf commented:
Reviewed 20 of 40 files at r1.
Review status: 19 of 36 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iJLM-dz3wFctNLJR:-K6_iJLNwHlpIDfPtfBA:-1198130620) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L327)):*
didn't we have another place where campaigning was triggered? Do we need this place still?

---

*[storage/replica.go, line 963 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_jHRVNfYBgAEJD9f7:-K6_jHRWj7RBa2ms_iwK:-1751087978) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L963)):*
not for this PR, but can we make `raftCommandIDLen` exactly the right size for `EncodeUint64` now?

---

*[storage/replica.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iszb7RaK4aHkjk5N:-K6_iszb7RaK4aHkjk5O:-2004015958) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L977)):*
`idKey` is purely random, so instead of exploding here, shouldn't we just randomize again?
Also, if we fail, it should be `ReplicaCorruptionError`.

---

*[storage/replica.go, line 983 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_jTrz4fo6uraIGr22:-K6_jTrz4fo6uraIGr23:-1032331) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L983)):*
slightly non-idiomatic to return a non-nil value and an error. Here you could give `proposePendingCmdLocked` the signature `(*pendingCmd, error)` to avoid that, though that's only slightly less non-idiomatic.

Also seems that on error, you'd want to clear the entry from the `pending` map.

---

*[storage/replica.go, line 991 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_jvtiJFsFeBeHX4e7:-K6_jvtjvpVMAXo5_ITW:-708341476) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L991)):*
`return r.proposeRaft.....`

---

*[storage/replica.go, line 998 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_k1ww9XfvRfV_Bi-R:-K6_k1ww9XfvRfV_Bi-S:-1536635884) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L998)):*
not having looked at `checkRaftGroup`, I have no idea what this does. A comment perhaps?

---

*[storage/replica.go, line 1002 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kAc_T09T0KntovUu:-K6_kAc_T09T0KntovUv:-1143483027) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1002)):*
`if !ok { continue }` saves a lot of indentation and looks nicer generally.

---

*[storage/replica.go, line 1045 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kQdsTQ5f7ST7WXdt:-K6_kQdsTQ5f7ST7WXdu:-2001498686) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1045)):*
no coalesced heartbeats, no TODO?

---

*[storage/replica.go, line 1066 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kZAqzR2SNOtJxPw7:-K6_kZAqzR2SNOtJxPw8:2063464470) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1066)):*
nit: s/In/in/

---

*[storage/replica.go, line 1108 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kiT5Xjbd6uMJaGp4:-K6_kiT5Xjbd6uMJaGp5:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1108)):*
ditto.

---

*[storage/replica.go, line 1114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_knnWQr94_EW2Ottq:-K6_knnWQr94_EW2Ottr:1596925420) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1114)):*
add a comment about this special case.

---

*[storage/replica_raftstorage.go, line 560 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_lLlvq7s4SwgVxFJt:-K6_lLlvq7s4SwgVxFJu:-2124039929) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_raftstorage.go#L560)):*
16 bytes seems stale

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 20 of 40 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 1521 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bfkyQCMSizio2Hbjy:-K6bfkyQCMSizio2Hbjz:-1717811956) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1521)):*
should there be a log line here? I'm assuming you don't want to return an error.

---

*[storage/store.go, line 1552 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bgCv6oKRdMovJPXPC:-K6bgCv7vYy9JvESrBMs:1613471167) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1552)):*
can't everything change between releasing the store lock and acquiring the replica lock?

---

*[storage/store.go, line 1574 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bg_NbgWlgZ_73AHzx:-K6bg_NbgWlgZ_73AHzy:1028748254) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1574)):*
what's the point of having the goroutine sit here?

---

*[storage/store.go, line 1587 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bghDXxp13GQhDjW66:-K6bghDXxp13GQhDjW67:731232290) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1587)):*
`defer ticker.Stop()`

---

*[storage/store.go, line 1590 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bhQJCzodHMBr_mbZq:-K6bhQJD_GXT3vzjvc-m:-403582464) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1590)):*
it's interesting that you acquire a read lock only but modify `r.pendingReplicas` below. Intentional?

---

*[storage/store.go, line 1641 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bh4sKFCrYH6XMeKL_:-K6bh4sKFCrYH6XMeKLa:-2094883193) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1641)):*
`getOrCreateReplicaLocked`?

---

*[storage/store.go, line 1643 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6biFav6oqZmJ4iL4NX:-K6biFav6oqZmJ4iL4NY:-1529994429) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1643)):*
looks like

```go
if ok {
	if err := r.setReplicaID(replicaID); err != nil {
		return nil, err
	}
	return r, nil
}
// the stuff in the branch below
```

is more legible.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Might want to search for all instances of `multiraft`.  I'd bet there are a few others still hidden in comments.

---

Reviewed 40 of 40 files at r1.
Review status: all files reviewed at latest revision, 37 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 192 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6blR2QA8gV7cJeIsT6:-K6blR2QA8gV7cJeIsT7:75522714) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L192)):*
This will be VERY useful.

---

*[storage/replica.go, line 295 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6blripm0eJHuIpygnX:-K6blripm0eJHuIpygnY:-734481419) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L295)):*
Would it makes sense to put the appliedIndex as a value protected by the replica lock instead of using an atomic?  Actually, appliedIndex, lastIndex, lease, desc, etc might also be candidates to use the replica lock instead. 

---

*[storage/replica.go, line 314 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dKpIGKIveG7k6NMpU:-K6dKpIGKIveG7k6NMpV:228398064) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L314)):*
You jump between `master` and `leader`. 

---

*[storage/replica.go, line 987 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dLYwLCEjio_MBYWTC:-K6dLYwMB9ccZR4EH5G3:-749950090) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L987)):*
There's a good amount of locking  and unlocking the replica lock in this function.  Does it makes sense to just lock at the top and defer the unlock?

---

*[storage/replica.go, line 1030 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMDoOD8-qMiOZ0jMe:-K6dMDoPbbBt8z_QrK5s:-71452555) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1030)):*
This function also has a ton of lock and unlocks in it.  having just the one would simplify it a lot.

---

*[storage/replica.go, line 1105 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dLvSX9KAM0TfmBzTa:-K6dLvSX9KAM0TfmBzTb:739854209) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1105)):*
can the store locking part be a function on the store?

---

*[storage/replica.go, line 1137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMXgJEmnOYm4jbh_T:-K6dMXgK6Oq9Dwuh9jX1:-159987127) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1137)):*
More locking, we can always pull back some of the locking later.

---

*[storage/replica.go, line 1147 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMwYD6Dgmzc87XcUX:-K6dMwYD6Dgmzc87XcUY:-1914101382) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1147)):*
can you move the store locking to a function on the store?

---

*[storage/replica_test.go, line 285 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bT3J5tOX6bIDmmH-J:-K6bSZdAHIJHKzXNRbXP:-1583164428) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L285)):*
nit: you're combining two different errors here, perhaps it would be best to have different codepaths for both.

---

*[storage/replica_test.go, line 732 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bTLmjePYdIkn6rC8a:-K6bTLmjePYdIkn6rC8b:-1271472888) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L732)):*
this one is a little worse, can proposeraftcommand proce a replica not found, or just the command itself?

---

*[storage/store.go, line 863 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6b_0NzxqyjWNSRAs6k:-K6b_0NzxqyjWNSRAs6l:-596109268) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L863)):*
Perhaps move the inner working of this to replica?  It seems strange to request the replica lock inside store unless there is a real need to do so.

---

*[storage/store.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6b_YSQOQidHLyAJFcE:-K6b_YSQOQidHLyAJFcF:1012731376) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd80624707adec8c7e23ae692b638383b09d0eae/storage/store.go#L977)):*
I love that you get to remove this line.  :)

---

*[storage/store.go, line 1512 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bb_cR45sZbcaD4pdp:-K6bb_cR45sZbcaD4pdq:-1061883917) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1512)):*
Any reason not to just hold the store lock for the whole function?

---

*[storage/store.go, line 1555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bbshq-4aCA2P_9Eae:-K6bbshreJksyglrm0gt:2083720370) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1555)):*
Could you move this to a func on replica?  So all functions that require a replica lock are on replica?

---

*[storage/store.go, line 1568 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bcfaes07V62Y10ltf:-K6bcfaes07V62Y10ltg:1140815824) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1568)):*
Why make this async?  

---

*[storage/store.go, line 1624 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bd0vHoKrBU_QTxTtc:-K6bd0vHoKrBU_QTxTtd:759419882) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1624)):*
again, move this replica locking to replica please.

---

*[storage/store.go, line 1784 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bdfbrhQ2vvmjjjEKi:-K6bdfbrhQ2vvmjjjEKj:-671715292) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1784)):*
fix this whole comment

---

*[storage/store.go, line 1790 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bdT_yvMtM-6_gTZwp:-K6bdT_yvMtM-6_gTZwq:-846436289) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1790)):*
indeed, this looks ugly.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-:-K6bfQ1jgaTy9b0c9oFp:-620028296)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 33 of 36 files reviewed at latest revision, 25 unresolved discussions.

---

*[storage/replica.go, line 295 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6blripm0eJHuIpygnX:-K6j7MUHEDz4vmTMDpoM:1083944841) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L295)):*
Yes, but that's a change for another day (#3270).

---

*[storage/replica.go, line 314 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dKpIGKIveG7k6NMpU:-K6j7cjNfLI-Y8LKn8bV:1758359711) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L314)):*
Fixed.

---

*[storage/replica.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iJLM-dz3wFctNLJR:-K6j7j4ERui3Lu8PHxC-:-1119005998) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L327)):*
Yes, in the split trigger. Both are still necessary.

---

*[storage/replica.go, line 963 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_jHRVNfYBgAEJD9f7:-K6j8EAIu-J0azHEdsSf:-1231522539) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L963)):*
`raftCommandIDLen` is already 8; I think the slicing here is vestigial. Removed.

---

*[storage/replica.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iszb7RaK4aHkjk5N:-K6j8olpyOsMOWB-3bKN:1346333618) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L977)):*
Collisions in 64 bits of randomness are too rare to be worth retrying. I've also added a TODO to revisit the command ID; I'm not sure if we need it anymore.

---

*[storage/replica.go, line 983 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_jTrz4fo6uraIGr22:-K6j9poP4AhJraCMavhG:1830405297) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L983)):*
Good point about removing the entry from `pending` on error; with that change we no longer return non-nil values on error.

---

*[storage/replica.go, line 987 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dLYwLCEjio_MBYWTC:-K6jAhHCQxeNrldzQPyc:-1703860009) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L987)):*
Was this comment intended for a different method? This one (`proposePendingCmdLocked`) doesn't touch the lock because the caller already holds it.

---

*[storage/replica.go, line 998 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_k1ww9XfvRfV_Bi-R:-K6jB7toBIHWcpxmO7PC:1727296185) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L998)):*
If we add a comment here, wouldn't the same logic lead us to place the same comment at every call to `checkRaftGroup`? It might be better to just rename the method to make it more self-explanatory. Something like `enqueueRaftUpdateCheck`?

---

*[storage/replica.go, line 1030 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMDoOD8-qMiOZ0jMe:-K6jBrGRpdvCzEDje2Zz:-477361537) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1030)):*
 I started with one function-scoped lock but there were deadlocks. Once the big PR is in I want to take another look at how all this is synchronized.

---

*[storage/replica.go, line 1045 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kQdsTQ5f7ST7WXdt:-K6jCO_J3toDR6oE9EGM:-1091077056) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1045)):*
Right, if we decide not to bring back coalesced heartbeats then this and other similar TODOs disappear.

---

*[storage/replica.go, line 1066 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_kZAqzR2SNOtJxPw7:-K6jCamU2W5rdNYvO0fA:420251747) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1066)):*
The parenthetical is multiple complete sentences; capitalizing "In" is appropriate.

---

*[storage/replica.go, line 1105 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dLvSX9KAM0TfmBzTa:-K6jClb01cpaf65dUQ7S:1261417239) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1105)):*
Yeah. Renamed `cacheReplicaDescriptor` and `replicaDescriptor` to add `Locked` suffix and added `cacheReplicaDescriptor` to grab the lock and call the locked version.

---

*[storage/replica.go, line 1137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMXgJEmnOYm4jbh_T:-K6jDYzOpz7YNX7o7nYD:1299520726) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1137)):*
The point of the comment is that we can't simply hold `r.Lock()` throughout this method because it deadlocks. If we need more locking it will have to involve new locks or other refactoring.

---

*[storage/replica.go, line 1147 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMwYD6Dgmzc87XcUX:-K6jDshuCV7dc6oq0U0I:140907985) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1147)):*
We could but then the two calls to `replicaDescriptorLocked` would be under two separate locks, which is not a good pattern to fall into (even though it's fine in this case).

---

*[storage/replica_test.go, line 285 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bT3J5tOX6bIDmmH-J:-K6jEKYL9nHgZNw-Ueq4:-1297565558) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L285)):*
The old code did the same; I'm just trying to minimize unnecessary changes in this PR.

---

*[storage/replica_test.go, line 732 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bTLmjePYdIkn6rC8a:-K6jEd2TjKSM-4KM4DhY:2018701157) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L732)):*
Again, the old code had the same ambiguity. 

---

*[storage/store.go, line 1512 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bb_cR45sZbcaD4pdp:-K6jG_87OUx-MgjQv0BX:-1881733291) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1512)):*
The lock must not be held when calling `raftGroup.Step`.

---

*[storage/store.go, line 1521 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bfkyQCMSizio2Hbjy:-K6jGlWDURRb0b8qVljM:491295879) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1521)):*
The old code didn't have a log line here.

---

*[storage/store.go, line 1552 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bgCv6oKRdMovJPXPC:-K6jGybMEVWjFGWDJEtX:-885269679) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1552)):*
Yeah, I'm not sure what the implications of that are. Added a TODO.

---

*[storage/store.go, line 1555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bbshq-4aCA2P_9Eae:-K6jHkKgyoQ0e3bZNiBm:1374363858) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1555)):*
I don't think that should be a goal: internal locking makes it more difficult to do multiple operations under the same lock, and I suspect we're going to need to do more of that in the future. 

---

*[storage/store.go, line 1568 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bcfaes07V62Y10ltf:-K6jMslpEgoxIJ02Xcac:598331007) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1568)):*
It may be called with `Replica.RWMutex` held; trying to acquire `s.mu` directly violates lock ordering and may deadlock (comment added). Spawning a goroutine was the most expedient solution, but is probably not what we really want, hence the TODO.

---

*[storage/store.go, line 1574 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bg_NbgWlgZ_73AHzx:-K6jN8DamfqUQDexH23f:1744631921) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1574)):*
Hmm, good question. Maybe the channel was unbuffered at some earlier point, but now it should be fine to do a non-blocking send and return immediately if the channel is full.

---

*[storage/store.go, line 1590 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bhQJCzodHMBr_mbZq:-K6jNiwtxfdKp3I-kAlY:668882738) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1590)):*
No. Fixed.

---

*[storage/store.go, line 1784 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bdfbrhQ2vvmjjjEKi:-K6jQIhlEXsBprb8ZmTV:-676471680) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1784)):*
Removed this section; we can do everything under s.mu now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 295 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6blripm0eJHuIpygnX:-K6jXObIACx8dYdfY3JS:-291415481) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L295)):*
Great.  I might just try to tackle that once this change is in.

---

*[storage/replica.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iszb7RaK4aHkjk5N:-K6jXdaBUQlWwDSmiyGO:-1416869987) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L977)):*
I see no reason why we can't remove the IDs entirely.  Just not worth doing in this PR.

---

*[storage/replica.go, line 987 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dLYwLCEjio_MBYWTC:-K6jY-FoONY2nSXKL2lz:-1522299956) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L987)):*
hmm I think it was for handleRaftReady(), but the comments there too.  Strange.  Ok, just ignore this one.

---

*[storage/replica.go, line 1030 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMDoOD8-qMiOZ0jMe:-K6jYCzAGWAuh_Q6McbB:-470515680) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1030)):*
cool.

---

*[storage/replica.go, line 1147 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6dMwYD6Dgmzc87XcUX:-K6jZ3Q-nuRiGlylK5-v:2007572405) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L1147)):*
Gotcha.  We could add a replicaDescriptorSLocked, but you're right, that would be strange.

---

*[storage/replica_test.go, line 285 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bT3J5tOX6bIDmmH-J:-K6jZGkJ_vOcpQWwHvj1:2062440116) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L285)):*
I noticed that.  Perhaps add a todo to clean this up.  It's really ugly.

---

*[storage/replica_test.go, line 732 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bTLmjePYdIkn6rC8a:-K6jZLTgsZG8TNfgrgol:-149221371) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica_test.go#L732)):*
same here

---

*[storage/store.go, line 1555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bbshq-4aCA2P_9Eae:-K6jZtFO9gvZWhL3W_b-:-1655651139) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1555)):*
For now, why not just make the raftGroup functions that are needed directly on the replica itself, and have them acquire and release the lock there.  If we need to run multiple raft functions, we can adjust accordingly ny either bundling them together in common functions on replica, or adding "Locked" versions of them.  

---

*[storage/store.go, line 1784 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bdfbrhQ2vvmjjjEKi:-K6j_ZLfOY2cgjIwl1By:5172980) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1784)):*
:)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
Avoid caching temporary (uncommitted) ReplicaDescriptors.

This fixes a timing-dependent failure in
TestChangeReplicasDescriptorInvariant. The bug has been present since
before the removal of multiraft, but the necessary timing conditions
were not observed until now.

Remove Replica.pendingReplica. It does not appear to be necessary any
more (if it ever was), and its presence invited bugs like the incorrect
caching of temporary values.

BramGruneir commented:
Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I scheduled 10 builds (and will schedule some more when they're done since I'm the only one awake anyway). There's obviously a lot of attack surface to work on the locking, so as long as it's stable now it LGTM. Had you investigated the failure TestChangeReplicasDescriptorInvariant in one of your previous builds?

---

Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[storage/replica.go, line 998 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_k1ww9XfvRfV_Bi-R:-K6mBkPd-d1K9OxdhFYm:1007110404) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L998)):*
sgtm, that gives a good idea.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-:-K6mDA-_qt76bqHOoKGM:1649931651)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
failed tests:
- TestAcquireLeaderLease
- leaked store in `ts` package
https://circleci.com/gh/cockroachdb/cockroach/10429
- RangeLookup dispatched to correct range, but no matching RangeDescriptor
was found: /Meta1/"a" [recovered]
https://circleci.com/gh/cockroachdb/cockroach/10431

nothing new there. I'll do 10 more to make sure.

On Wed, Dec 30, 2015 at 12:57 AM Bram Gruneir <notifications@github.com>
wrote:

> Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
> Review status: all files reviewed at latest revision, 25 unresolved
> discussions.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3528>*
>
> —
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3528#issuecomment-167942012>
> .
>

tschottdorf commented:
TestConvergence https://circleci.com/gh/cockroachdb/cockroach/10433
`ts` leaking store again https://circleci.com/gh/cockroachdb/cockroach/10436
https://circleci.com/gh/cockroachdb/cockroach/10439
TestChangeReplicasDuplicateError
https://circleci.com/gh/cockroachdb/cockroach/10438
TestMerge (interesting but unrelated):
https://circleci.com/gh/cockroachdb/cockroach/10438, filed #3555

looks like Raft-wise, this is doing fine (or at least not worse than
before).

On Wed, Dec 30, 2015 at 6:40 AM Tobias Schottdorf <tobias@cockroachlabs.com>
wrote:

> failed tests:
> - TestAcquireLeaderLease
> - leaked store in `ts` package
> https://circleci.com/gh/cockroachdb/cockroach/10429
> - RangeLookup dispatched to correct range, but no matching RangeDescriptor
> was found: /Meta1/"a" [recovered]
> https://circleci.com/gh/cockroachdb/cockroach/10431
>
> nothing new there. I'll do 10 more to make sure.
>
> On Wed, Dec 30, 2015 at 12:57 AM Bram Gruneir <notifications@github.com>
> wrote:
>
>> Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
>> Review status: all files reviewed at latest revision, 25 unresolved
>> discussions.
>> ------------------------------
>>
>> *Comments from the review on Reviewable.io
>> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3528>*
>>
>> —
>>
>>
>> Reply to this email directly or view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/3528#issuecomment-167942012>
>> .
>>
>

bdarnell commented:
`TestChangeReplicasDescriptorInvariant` is fixed in the latest commits in this branch. I just filed #3561 for `TestAcquireLeaderLease`. The `ts` store leak is not new to this branch; Andrei saw it yesterday. I'm going to spend some more time investigating all of the failures since I've seen some non-obvious failures caused by race conditions while working on this branch.

JackKrupansky commented:
I see two lingering references to Multiraft, both undo TODOs:

in server/raft_transport.go:
// TODO(bdarnell): consider changing raftServiceName/raftMessageName
  	raftServiceName = "MultiRaft"
  	raftMessageName = raftServiceName + ".RaftMessage"

in storage/raft_transport.go:
// TODO(bdarnell): remove/change raftMessageName
 const raftMessageName = "MultiRaft.RaftMessage"

JackKrupansky commented:
The CockroachDB design document doesn't refer to Multiraft by name, but does say "Our Raft implementation was developed together with CoreOS, but adds an extra layer of optimization to account for the fact that a single Node may have millions of consensus groups (one for each Range). Areas of optimization are chiefly coalesced heartbeats (so that the number of nodes dictates the number of heartbeats as opposed to the much larger number of ranges) and batch processing of requests."

Is that still an approximately accurate summary of the optimization after this pull request, or has it changed in some major way?

Is it still "Raft" that is doing the optimization, or should we speak of some other module (storage or whatever) that is optimizing for a large number of Raft ranges?

BramGruneir commented:
@JackKrupansky 
The lingering multiraft references are fine.  Since they have a todo next to them that we'll fix up with a new name shortly (or possibly keep it the same).
This change is about moving the multiraft functionality out of it's own package and onto the storage package directly and does not change the overall design. We found that we were running into hard to debug and detect concurrency issues with the extra level of abstraction.  See [dismantle multiraft](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/dismantle_multiraft.md) for more details on the motivations behind it.  
The two main optimizations, Coalesced Heartbeats and Batched Processing will still remain, just no longer abstracted.  (Coalesced heartbeats will reappear shortly in a follow up PR.)

es-chow commented on storage/replica.go:
```diff
@@ -132,8 +133,10 @@ func usesTimestampCache(r roachpb.Request) bool {
-	ctx  context.Context
-	done chan roachpb.ResponseWithError // Used to signal waiting RPC handler
+	ctx     context.Context
+	idKey   cmdIDKey
```
idKey can be removed if change signature of proposePendingCmdLocked to take idKey as input parameter.

es-chow commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 1574 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6s330SM7BQBFnT8ogU:-K6s330SM7BQBFnT8ogV:1621043148) ([raw file](https://github.com/cockroachdb/cockroach/blob/663f181232801f0e175a84ea97350c7086c51d82/storage/store.go#L1574)):*
This goroutine and lock can be replaced by channel of rangeID. In `processRaft`, the length of same channel can be checked, if length is great than 0, then make a rangeID map from the length elements from the channel.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

es-chow commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 1617 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6s8VdZj6QXOD4OG-Sg:-K6s8VdZj6QXOD4OG-Sh:2126907437) ([raw file](https://github.com/cockroachdb/cockroach/blob/663f181232801f0e175a84ea97350c7086c51d82/storage/store.go#L1617)):*
Can this channel also be removed?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
Fix a lock leak on error in ChangeReplicas.

bdarnell commented:
I've fixed `TestChangeReplicasDuplicateError`; the remaining test failures are also present on `master` so I think this is ready to go (although this change did aggravate the `TestMerge` failure as described in #3555). 

---

Review status: 35 of 36 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iszb7RaK4aHkjk5N:-K6t9wLeaHdKx_vNzCxZ:-232926815) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L977)):*
How else would we associate a committed entry with its waiting `done` channel?

---

*[storage/store.go, line 1555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6bbshq-4aCA2P_9Eae:-K6tBliwA_crAFk_spgc:540778837) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/store.go#L1555)):*
That just seems like a lot of busy work for no practical benefit. Tightly-coupled raft logic is present in `store.go` whether it accesses `Replica.raftGroup` directly or goes through a thin wrapper method.

---

*[storage/store.go, line 1574 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6s330SM7BQBFnT8ogU:-K6tCx18qV6HTVtY6G__:-560419429) ([raw file](https://github.com/cockroachdb/cockroach/blob/663f181232801f0e175a84ea97350c7086c51d82/storage/store.go#L1574)):*
We could avoid the lock by passing range IDs over the channel but then there is the possibility of deadlock when the channel gets full. There are ways around that but they add complexity, so I'm going to look into this later.

---

*[storage/store.go, line 1617 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6s8VdZj6QXOD4OG-Sg:-K6tDGYhW5r5hD-cH9kV:-1739426506) ([raw file](https://github.com/cockroachdb/cockroach/blob/663f181232801f0e175a84ea97350c7086c51d82/storage/store.go#L1617)):*
I don't think we can remove it yet: the fact that `removeReplicaImpl` and `handleRaftReady` are both called from the `processRaft` goroutine is the reason we know we won't remove a replica while it is working. However, in a future change I'd like to move this to a lock instead of a single goroutine (this is a prerequisite for parallelizing raft commands as in #3244).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-:-K6tEW_J__hSJGNk3P-n:1770025085)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 4 of 8 files at r5.
Review status: 32 of 36 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/replica.go, line 977 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528#-K6_iszb7RaK4aHkjk5N:-K6uY8It55ZUYahn2ogX:1748494638) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ccc9945681c4d72a39e44a0164f2b7184bfbf9/storage/replica.go#L977)):*
Does pendingCmds need to be a map?  Couldn't it be a queue instead?  But this is a discussion for another time.  

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 4 of 8 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3528)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
LGTM
